---
title: "Differential Gene Expression Analysis - TCGA"
output: html_notebook
---

## I. Preliminaries

### Loading libraries

```{r}
library('tidyverse')
library('tibble')
library('msigdbr')
library('ggplot2')
library('TCGAbiolinks')
library('RNAseqQC')
library('DESeq2')
library('ensembldb')
library('purrr')
library('magrittr')
library('vsn')
library('matrixStats')
library('dplyr') 
library('grex')
```

## II. Downloading the TCGA gene expression data 

Download gene expression data from The Cancer Genome Atlas (TCGA):
- `TCGA-COAD` refers to the biospecimen data for colon adenocarcinoma.
- `STAR - Counts` pertains to the raw counts.

```{r}
query_tumor <- GDCquery(
  project = "TCGA-COAD",
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification",
  experimental.strategy = "RNA-Seq",
  workflow.type = "STAR - Counts",
  access = "open",
  sample.type = "Primary Tumor"
)
tumor <- getResults(query_tumor)
tumor
```

```{r}
query_normal <- GDCquery(
  project = "TCGA-COAD",
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification",
  experimental.strategy = "RNA-Seq",
  workflow.type = "STAR - Counts",
  access = "open",
  sample.type = "Solid Tissue Normal"
)
normal <- getResults(query_normal)
normal
```
Consider only samples with both normal and malignant tissues.

```{r}
submitter_ids <- inner_join(tumor, normal, by = "cases.submitter_id") %>%
              dplyr::select(cases.submitter_id)
tumor <- tumor %>%
  dplyr::filter(cases.submitter_id %in% submitter_ids$cases.submitter_id)
normal <- normal %>%
  dplyr::filter(cases.submitter_id %in% submitter_ids$cases.submitter_id)

samples <- rbind(tumor, normal)
unique(samples$sample_type)
samples
```

Download only samples with both normal and malignant tissues.

To impose this filtering, we set the `barcode` argument of `GDCquery` to `samples$sample.submitter_id` (which was generated in the previous cell).

```{r}
query_coad <- GDCquery(
  project = "TCGA-COAD",
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification",
  experimental.strategy = "RNA-Seq",
  workflow.type = "STAR - Counts",
  access = "open",
  sample.type = c("Solid Tissue Normal", "Primary Tumor"),
  barcode = as.list(samples$sample.submitter_id)
)
```
If this is your first time running this notebook (i.e., you have not yet downloaded the results of the query in the previous block), uncomment the code block below.

```{r}
# GDCdownload(query_coad)
```

Running the code block above should generate and populate a directory named `GDCdata`.

## III. Data preprocessing

Display the RNA-seq count matrix.

```{r}
tcga_coad_data <- GDCprepare(query_coad, summarizedExperiment = TRUE)
count_matrix <- assay(tcga_coad_data, 'unstranded')
rownames(count_matrix) <- cleanid(rownames(count_matrix))
head(count_matrix)
```
```{r}
rownames(samples) <- samples$cases
samples <- samples %>%
  dplyr::select(case="cases.submitter_id", type="sample_type")
samples$type <- str_replace(samples$type, "Solid Tissue Normal", "normal")
samples$type <- str_replace(samples$type, "Primary Tumor", "tumor")
```

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts_matrix,
                              colData = samples,
                              design = ~ type)
dds
```