---
title: "GTEx Data Analysis Notebook"
output: html_notebook
---

## **Loading libraries**

```{r}
library('tidyverse')
library('tibble')
library('msigdbr')
library('ggplot2')
library('TCGAbiolinks')
library('RNAseqQC')
library('DESeq2')
library('ensembldb')
library('purrr')
library('magrittr')
library('dplyr') 
library('vsn')
library('matrixStats')
```

## **Loading the GTEx annotations**

```{r}
sample.df <- read.delim("data/GTEx/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt", as.is=TRUE, header=TRUE, row.names=1)
subject.df <- read.delim("data/GTEx/GTEx_Analysis_v8_Annotations_SubjectPhenotypesDS.txt", as.is=TRUE, header=TRUE, row.names=1)
```

```{r}
head(subject.df)
```

```{r}
head(sample.df)
```

```{r}
rnaseq.sample.df <- sample.df[sample.df['SMAFRZE']=='RNASEQ', ]
```

```{r}
head(rnaseq.sample.df)
```

## **Filtering Colon samples**

```{r}
as.matrix(sort(table(rnaseq.sample.df['SMTSD']), decreasing = TRUE))
```

```{r}
colon.sample.df <- rnaseq.sample.df  %>% stats::filter('SMTSD' == 'Colon - Sigmoid')
colon.sample.df
```

## **Loading GTEx data**

```{r}
tpm.df <- read.delim('data/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct',
                     as.is=T, row.names=1, check.names=FALSE, skip=2)
gene.names.df <- tpm.df[, 'Description', drop=FALSE]
tpm.df <- tpm.df[, !(names(tpm.df) %in% c('Description'))]
{cat(paste("Number of genes in table: ", dim(tpm.df)[1]))}
```

```{r}
tpm.copy.df <- tpm.df
```

```{r}
tpm.df$ensembl <- cleanid(rownames(tpm.df))
head(tpm.df$ensembl)
```

```{r}
head(tpm.df)
```

```{r}
# Create a named vector to map names to IDs
name_to_id <- setNames(rownames(gene.names.df), gene.names.df$Description)

# Create a named vector to map IDs to names
id_to_name <- setNames(gene.names.df$Description, rownames(gene.names.df))

# To retrieve the ID for 'FADD'
#name_to_id[['FADD']]
id_to_name[['ENSG00000168040.4']]
```

## **Loading RCD gene sets**

*Necroptosis*

```{r}
necroptosis.genes <- msigdbr(species = "human", category = "C5", subcategory = "GO:BP")  %>%
  filter(gs_name == "GOBP_NECROPTOTIC_SIGNALING_PATHWAY")
head(necroptosis.genes)
```

*Ferroptosis*

```{r}
ferroptosis.genes <- msigdbr(species = "human", category = "C2", subcategory = "CP:WIKIPATHWAYS")  %>%
  filter(gs_name == "WP_FERROPTOSIS")
head(ferroptosis.genes)
```

*Pyroptosis*

```{r}
pyroptosis.genes <- msigdbr(species = "human", category = "C2", subcategory = "CP:REACTOME")  %>%
  filter(gs_name == "REACTOME_PYROPTOSIS")
head(pyroptosis.genes)
```

## Exploratory Data Analysis

```{r}
tpm.colon.df <- tpm.df %>% select(c(ensembl, rownames(colon.sample.df)))
tpm.colon.df
```

### Necroptosis

```{r}
tpm.colon.necro.df <- tpm.colon.df %>% filter(ensembl %in% necroptosis.genes$ensembl_gene)
tpm.colon.necro.df

tpm.colon.necro.df2 <- left_join(tpm.colon.necro.df, necroptosis.genes %>% select(ensembl_gene, gene_symbol), by = c("ensembl"="ensembl_gene"))
tpm.colon.necro.df$gene_symbol <- tpm.colon.necro.df2$gene_symbol
tpm.colon.necro.df

results.samples.necro <- tpm.colon.necro.df %>% select(-c('ensembl', 'gene_symbol'))
results.samples.necro
```

```{r}
gene.expressions.necro <- data.frame (
  row.names = tpm.colon.necro.df$gene_symbol,
  tissue = 'Colon',
  tpm = matrixStats::rowMedians(as.matrix(tpm.colon.necro.df%>%select(-c('ensembl', 'gene_symbol'))))
)
gene.expressions.necro
```

#### Gene expression plot

```{r}
ggplot(gene.expressions.necro, aes(y = tissue, x = rownames(gene.expressions.necro), fill = tpm)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  theme_minimal() +
  theme(
    axis.text.x = element_text(size = 9, angle = 90, hjust = 1)
  ) +
  labs(
    title = "Expression of Necroptosis-related regulators in GTEx Colon Tissues",
    x = "Tissue",
    y = "GeneName",
    fill = "TPM"
  ) +
  geom_text(aes(label = tpm), vjust = 1)
```

### Ferroptosis

```{r}
tpm.colon.ferro.df <- tpm.colon.df %>% filter(ensembl %in% ferroptosis.genes$ensembl_gene)
tpm.colon.ferro.df

tpm.colon.ferro.df2 <- left_join(tpm.colon.ferro.df, ferroptosis.genes %>% select(ensembl_gene, gene_symbol), by = c("ensembl"="ensembl_gene"))
tpm.colon.ferro.df$gene_symbol <- tpm.colon.ferro.df2$gene_symbol
tpm.colon.ferro.df

results.samples.ferro <- tpm.colon.ferro.df %>% select(-c('ensembl', 'gene_symbol'))
results.samples.ferro
```

```{r}
gene.expressions.ferro <- data.frame (
  row.names = tpm.colon.ferro.df$gene_symbol,
  tissue = 'Colon',
  tpm = matrixStats::rowMedians(as.matrix(tpm.colon.ferro.df%>%select(-c('ensembl', 'gene_symbol'))))
)
gene.expressions.ferro
```

#### Gene expression plot

```{r}
ggplot(gene.expressions.ferro, aes(y = tissue, x = rownames(gene.expressions.ferro), fill = tpm)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  theme_minimal() +
  theme(
    axis.text.x = element_text(size = 7, angle = 90, hjust = 1)
  ) +
  labs(
    title = "Expression of Ferroptosis-related regulators in GTEx Colon Tissues",
    x = "Tissue",
    y = "GeneName",
    fill = "TPM"
  )
```

###  Pyroptosis

```{r}
tpm.colon.pyro.df <- tpm.colon.df %>% filter(ensembl %in% pyroptosis.genes$ensembl_gene)
tpm.colon.pyro.df

tpm.colon.pyro.df2 <- left_join(tpm.colon.pyro.df, pyroptosis.genes %>% select(ensembl_gene, gene_symbol), by = c("ensembl"="ensembl_gene"))
tpm.colon.pyro.df$gene_symbol <- tpm.colon.pyro.df2$gene_symbol
tpm.colon.pyro.df

results.samples.pyro <- tpm.colon.pyro.df %>% select(-c('ensembl', 'gene_symbol'))
results.samples.pyro
```

```{r}
gene.expressions.pyro <- data.frame (
  row.names = tpm.colon.pyro.df$gene_symbol,
  tissue = 'Colon',
  tpm = matrixStats::rowMedians(as.matrix(tpm.colon.pyro.df%>%select(-c('ensembl', 'gene_symbol'))))
)
gene.expressions.pyro
```

#### Gene expression plot

```{r}
ggplot(gene.expressions.pyro, aes(y = tissue, x = rownames(gene.expressions.pyro), fill = tpm)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  theme_minimal() +
  theme(
    axis.text.x = element_text(size = 8, angle = 90, hjust = 1)
  ) +
  labs(
    title = "Expression of Pyroptosis-related regulators in GTEx Colon Tissues",
    x = "Tissue",
    y = "GeneName",
    fill = "TPM"
  )
```

## **Differential expression analysis in TCGA datasets**

```{r}
# Querying All Colorectal Cancer Data (Tumor)
query_tumor <- GDCquery(
  project = "TCGA-COAD",
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification",
  experimental.strategy = "RNA-Seq",
  workflow.type = "STAR - Counts",
  access = "open",
  sample.type = "Primary Tumor"
)
tumor <- getResults(query_tumor)
```

```{r}
# Querying All Colorectal Cancer Data (Normal)
query_normal <- GDCquery(
  project = "TCGA-COAD",
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification",
  experimental.strategy = "RNA-Seq",
  workflow.type = "STAR - Counts",
  access = "open",
  sample.type = "Solid Tissue Normal"
)
normal <- getResults(query_normal)
```

```{r}
# Filtering samples with both tumor and normal tissues
submitter_ids <- inner_join(tumor, normal, by = "cases.submitter_id") %>%
              select(cases.submitter_id)
tumor <- tumor %>%
  filter(cases.submitter_id %in% submitter_ids$cases.submitter_id)
normal <- normal %>%
  filter(cases.submitter_id %in% submitter_ids$cases.submitter_id)

# Merging to a single data.frame
samples <- rbind(tumor, normal)
unique(samples$sample_type)
samples$sample.submitter_id
```

```{r}
# Overwrite barcode of query with the filtered sample ids
query_coad <- GDCquery(
  project = "TCGA-COAD",
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification",
  experimental.strategy = "RNA-Seq",
  workflow.type = "STAR - Counts",
  access = "open",
  sample.type = c("Solid Tissue Normal", "Primary Tumor"),
  barcode = as.list(samples$sample.submitter_id)
)
getResults(query_coad)

```

```{r}
# Downloading TCGA data
GDCdownload(query_coad)
```

### **Data wrangling for DE Analysis**

```{r}
# prepare count data
tcga_coad_data <- GDCprepare(query_coad, summarizedExperiment = TRUE)
head(tcga_coad_data)
#tcga_coad_data <- tcga_coad_data %>%
#                    filter()

# unstanded, stranded_first, stranded_second, tpm_unstrand, fpkm_unstrand, fpkm_uq_unstrand
coad_matrix <- assay(tcga_coad_data, 'unstranded')
head(coad_matrix)
```

```{r}
# make dataframe of sample names, cell Lines, and tissue type(normal or tumor)
rownames(samples) <- samples$cases
View(samples)
samples <- samples %>%
  select(case="cases.submitter_id", type="sample_type")
samples$type <- str_replace(samples$type, "Solid Tissue Normal", "normal")
samples$type <- str_replace(samples$type, "Primary Tumor", "tumor")

coad_matrix <- coad_matrix[, rownames(samples)]
# Check if all samples in the counts dataframe are in the samples dataframe
all(colnames(coad_matrix) %in% rownames((samples)))
all(colnames(coad_matrix) == rownames(samples))

a <- str_split_fixed(rownames(coad_matrix), "[.]", 2)
counts_matrix <- coad_matrix
rownames(counts_matrix) <- a[, 1]
head(counts_matrix)
```

### **Quality Control**

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts_matrix,
                              colData = samples,
                              design = ~ type)
```

#### QC plots on raw counts

```{r}
# QC plots on raw count matrix
plot_total_counts(dds)
plot_library_complexity(dds)
plot_gene_detection(dds)
```

```{r}
# Gene biotypes
plot_biotypes(dds)
```

```{r}
# Gene filtering
# filter to keep only rows that have 10 reads total. before: 45194 x 87 after: 53996 x 519
keep <- rowSums(counts(dds)) >= 10
dds.filtered <- dds[keep,]
```

#### Mean SD of transformed read counts

```{r}
# Variance stabilization
vsd <-vst(dds)
sdplot <-meanSdPlot(assay(vsd))
sdplot$gg <- sdplot$gg +
  ggtitle(label="Mean SD of transformed read counts") +
  ylab("standard deviation")
print(sdplot$gg)
```

```{r}
# Chromosomal expression
map(c("1", "5", "14"), ~plot_chromosome(vsd, .x))
```

```{r}
# Replicate variability
# define new grouping variable
colData(vsd)$trt_mut <- paste0(colData(vsd)$type, "_", colData(vsd)$case)

ma_plots <- plot_sample_MAs(vsd, group = "trt_mut")
cowplot::plot_grid(plotlist = ma_plots[17:24], ncol = 2)
```

#### Clustering

```{r}
# Clustering
# set seed to control random annotation colors
set.seed(1)
plot_sample_clustering(vsd, anno_vars = c("type"), distance = "euclidean")
```

#### Principal component analysis (PCA) plot

```{r}
# Principal component analysis (PCA) plot (from Tutorial)
plot_pca(vsd, PC_x = 1, PC_y = 2, color_by = "type")
```

```{r}
# PCA plot (from Ms Jenny)
pcaData<-plotPCA(vsd, intgroup=c("type"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=type, shape=type)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance-=")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```

### Necroptosis

#### **Filtering Necroptosis-related regulators**

```{r}
rownames(necroptosis.genes) <- necroptosis.genes$ensembl_gene
coad_necroptosis <- counts_matrix[rownames(counts_matrix) %in% necroptosis.genes$ensembl_gene, ]
coad_necroptosis <- coad_necroptosis[, rownames(samples)]

# Check if all samples in the counts dataframe are in the samples dataframe
all(colnames(coad_necroptosis) %in% rownames((samples)))
all(colnames(coad_necroptosis) == rownames(samples))
```

#### Differential Expression Analysis (Necroptosis)

```{r}
dds <- DESeqDataSetFromMatrix(countData = coad_necroptosis,
                              colData = samples,
                              design = ~ type)
dds
```

```{r}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
dds

dds$type <- relevel(dds$type, ref = "normal")

dds <- DESeq(dds)
res <- results(dds)
res
summary(res)

res0.05 <- results(dds, alpha = 0.05)
summary(res0.05)

resultsNames(dds)
```

#### Plot

```{r}
deseq.results <- res
deseq.bbl.data <- data.frame(
  row.names = rownames(deseq.results),
  baseMean = deseq.results$baseMean,
  log2FoldChange = deseq.results$log2FoldChange,
  lfcSE = deseq.results$lfcSE,
  stat = deseq.results$stat,
  pvalue = deseq.results$pvalue,
  padj = deseq.results$padj,
  cancer_type = 'Colon',
  gene_symbol = necroptosis.genes[rownames(deseq.results), 'gene_symbol']
)
deseq.bbl.data
```

```{r}
# “Genes with the threshold of FC>2 and FDR ≤ 0.05 were considered as significantly differentially expressed”
ggplot(deseq.bbl.data, aes(x=cancer_type, y=gene_symbol, size=padj, fill=log2FoldChange)) +
    geom_point(alpha=0.5, shape=21, color="black") +
    scale_size(trans='reverse') +
    scale_fill_gradient2(low = "blue", mid = "white", high = "red", limits = c(-1.5,3)) +
    theme_minimal() +
    theme(legend.position="bottom") +
    theme(legend.position = "bottom")+
    labs(size = "FDR", fill = "log2 FC", x = "Cancer type", y = "Gene")
```

### Ferroptosis

#### Filtering Ferroptosis-related regulators

```{r}
rownames(ferroptosis.genes) <- ferroptosis.genes$ensembl_gene
coad_ferroptosis <- counts_matrix[rownames(counts_matrix) %in% ferroptosis.genes$ensembl_gene, ]
coad_ferroptosis <- coad_ferroptosis[, rownames(samples)]

# Check if all samples in the counts dataframe are in the samples dataframe
all(colnames(coad_ferroptosis) %in% rownames((samples)))
all(colnames(coad_ferroptosis) == rownames(samples))
```

#### Differential Expression Analysis (Ferroptosis)

```{r}
dds <- DESeqDataSetFromMatrix(countData = coad_ferroptosis,
                              colData = samples,
                              design = ~ type)
dds
```

```{r}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
dds

dds$type <- relevel(dds$type, ref = "normal")

dds <- DESeq(dds)
res <- results(dds)
res
summary(res)

res0.05 <- results(dds, alpha = 0.05)
summary(res0.05)
```

#### Plot

```{r}
deseq.results <- res
deseq.bbl.data <- data.frame(
  row.names = rownames(deseq.results),
  baseMean = deseq.results$baseMean,
  log2FoldChange = deseq.results$log2FoldChange,
  lfcSE = deseq.results$lfcSE,
  stat = deseq.results$stat,
  pvalue = deseq.results$pvalue,
  padj = deseq.results$padj,
  cancer_type = 'Colon',
  gene_symbol = ferroptosis.genes[rownames(deseq.results), 'gene_symbol']
)
deseq.bbl.data
```

```{r}
# “Genes with the threshold of FC>2 and FDR ≤ 0.05 were considered as significantly differentially expressed”
ggplot(deseq.bbl.data, aes(x=cancer_type, y=gene_symbol, size=padj, fill=log2FoldChange)) +
    geom_point(alpha=0.5, shape=21, color="black") +
    scale_size(trans='reverse') +
    scale_fill_gradient2(low = "blue", mid = "white", high = "red", limits = c(-1.5,3)) +
    theme_minimal() +
    theme(legend.position="bottom") +
    theme(legend.position = "bottom")+
    labs(size = "FDR", fill = "log2 FC", x = "Cancer type", y = "Gene")
```

### Pyroptosis

#### Filtering Pyroptosis-related regulators

```{r}
rownames(pyroptosis.genes) <- pyroptosis.genes$ensembl_gene
coad_pyroptosis <- counts_matrix[rownames(counts_matrix) %in% pyroptosis.genes$ensembl_gene, ]
coad_pyroptosis <- coad_pyroptosis[, rownames(samples)]

# Check if all samples in the counts dataframe are in the samples dataframe
all(colnames(coad_pyroptosis) %in% rownames((samples)))
all(colnames(coad_pyroptosis) == rownames(samples))
```

#### Differential Expression Analysis

```{r}
dds <- DESeqDataSetFromMatrix(countData = coad_pyroptosis,
                              colData = samples,
                              design = ~ type)
dds
```

```{r}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
dds

dds$type <- relevel(dds$type, ref = "normal")

dds <- DESeq(dds)
res <- results(dds)
res
summary(res)

res0.05 <- results(dds, alpha = 0.05)
summary(res0.05)
```

#### Plot

```{r}
deseq.results <- res
deseq.bbl.data <- data.frame(
  row.names = rownames(deseq.results),
  baseMean = deseq.results$baseMean,
  log2FoldChange = deseq.results$log2FoldChange,
  lfcSE = deseq.results$lfcSE,
  stat = deseq.results$stat,
  pvalue = deseq.results$pvalue,
  padj = deseq.results$padj,
  cancer_type = 'Colon',
  gene_symbol = pyroptosis.genes[rownames(deseq.results), 'gene_symbol']
)
deseq.bbl.data
```

```{r}
# “Genes with the threshold of FC>2 and FDR ≤ 0.05 were considered as significantly differentially expressed”
ggplot(deseq.bbl.data, aes(x=cancer_type, y=gene_symbol, size=padj, fill=log2FoldChange)) +
    geom_point(alpha=0.5, shape=21, color="black") +
    scale_size(trans='reverse') +
    scale_fill_gradient2(low = "blue", mid = "white", high = "red", limits = c(-1.5,3)) +
    theme_minimal() +
    theme(legend.position="bottom") +
    theme(legend.position = "bottom")+
    labs(size = "FDR", fill = "log2 FC", x = "Cancer type", y = "Gene")
```
