---
title: "GTEx Colorectal Analysis Notebook"
output: html_notebook
---

## I. Preliminaries

### Loading libraries

```{r}
library('tidyverse')
library('tibble')
library('msigdbr')
library('ggplot2')
library('TCGAbiolinks')
library('RNAseqQC')
library('DESeq2')
library('ensembldb')
library('purrr')
library('magrittr')
library('vsn')
library('matrixStats')
library('dplyr') 
library('grex')
```

### Constants
```{r}
DATA_DIR <- "data/GTEx/"
```

## II. Loading the GTEx annotations

```{r}
sample.df <- read.delim(paste0(DATA_DIR, "GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt"), as.is=TRUE, header=TRUE, row.names=1)
subject.df <- read.delim(paste0(DATA_DIR, "GTEx_Analysis_v8_Annotations_SubjectPhenotypesDS.txt"), as.is=TRUE, header=TRUE, row.names=1)
```

```{r}
head(subject.df)
```

```{r}
head(sample.df)
```

```{r}
rnaseq.sample.df <- sample.df[sample.df['SMAFRZE']=='RNASEQ', ]
head(rnaseq.sample.df)
```

## III. Filtering Colon samples

```{r}
as.matrix(sort(table(rnaseq.sample.df['SMTSD']), decreasing = TRUE))
```

```{r}
colon.sample.df <- rnaseq.sample.df %>% dplyr::filter(SMTSD == "Colon - Sigmoid")
head(colon.sample.df)
```

## IV. Loading GTEx data

```{r}
tpm.df <- read.delim(paste0(DATA_DIR, "GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct"),
                     as.is=T, row.names=1, check.names=FALSE, skip=2)
gene.names.df <- tpm.df[, 'Description', drop=FALSE]
tpm.df <- tpm.df[, !(names(tpm.df) %in% c('Description'))]
{cat(paste("Number of genes in table: ", dim(tpm.df)[1]))}
```
```{r}
# Remove version number: https://www.rdocumentation.org/packages/grex/versions/1.9/topics/cleanid
tpm.df$ensembl <- cleanid(rownames(tpm.df))
head(tpm.df$ensembl)
```
```{r}
# Create a named vector to map names to IDs
name_to_id <- setNames(rownames(gene.names.df), gene.names.df$Description)

# Create a named vector to map IDs to names
id_to_name <- setNames(gene.names.df$Description, rownames(gene.names.df))

# To retrieve the ID for 'FADD'
#name_to_id[['FADD']]
id_to_name[['ENSG00000168040.4']]
```

## V. Loading RCD gene sets

### Necroptosis

```{r}
necroptosis.genes <- msigdbr(species = "human", category = "C5", subcategory = "GO:BP")  %>%
  dplyr::filter(gs_name == "GOBP_10_FORMYLTETRAHYDROFOLATE_METABOLIC_PROCESS")
head(necroptosis.genes)
```

### Ferroptosis

```{r}
ferroptosis.genes <- msigdbr(species = "human", category = "C2", subcategory = "CP:WIKIPATHWAYS")  %>%
  dplyr::filter(gs_name == "WP_FERROPTOSIS")
head(ferroptosis.genes)
```

### Pyroptosis

```{r}
pyroptosis.genes <- msigdbr(species = "human", category = "C2", subcategory = "CP:REACTOME")  %>%
  dplyr::filter(gs_name == "REACTOME_PYROPTOSIS")
head(pyroptosis.genes)
```

## VI. Exploratory Data Analysis

```{r}
tpm.colon.df <- tpm.df %>% dplyr::select(c(ensembl, rownames(colon.sample.df)))
tpm.colon.df
```

### Necroptosis

```{r}
tpm.colon.necro.df <- tpm.colon.df %>% dplyr::filter(ensembl %in% necroptosis.genes$ensembl_gene)
tpm.colon.necro.df

tpm.colon.necro.df2 <- left_join(tpm.colon.necro.df, necroptosis.genes %>% dplyr::select(ensembl_gene, gene_symbol), by = c("ensembl"="ensembl_gene"))
tpm.colon.necro.df$gene_symbol <- tpm.colon.necro.df2$gene_symbol
tpm.colon.necro.df

results.samples.necro <- tpm.colon.necro.df %>% dplyr::select(-c('ensembl', 'gene_symbol'))
results.samples.necro
```
```{r}
gene.expressions.necro <- data.frame (
  row.names = tpm.colon.necro.df$gene_symbol,
  tissue = 'Colon',
  tpm = matrixStats::rowMedians(as.matrix(tpm.colon.necro.df%>%dplyr::select(-c('ensembl', 'gene_symbol'))))
)
gene.expressions.necro
```
### Gene expression plot

```{r}
ggplot(gene.expressions.necro, aes(y = tissue, x = rownames(gene.expressions.necro), fill = tpm)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  theme_minimal() +
  theme(
    axis.text.x = element_text(size = 9, hjust = 1)
  ) +
  labs(
    title = "Expression of Necroptosis-related regulators in GTEx Colon Tissues",
    x = "Tissue",
    y = "GeneName",
    fill = "TPM"
  ) +
  geom_text(aes(label = tpm), vjust = 1)
```